dascore/io/xml_binary/utils.py

```diff
@@ -0,0 +1,233 @@
+    output_temporal_sampling_rate: float
+    original_spatial_sampling_interval: float
+    units: str
+    zones: dict[str, XMLLaserZones]
+    number_of_channels: int
+    number_of_frames: int
+    use_relative_strain: bool
+    transposed_data: bool
+
+
+@lru_cache
+def _read_xml_metadata(path):
+    """A function to read metadata from the xml file."""
+    contents = xml_to_dict(path.read_bytes())
+    return XMLBinaryInfo.model_validate(contents)
+
+
+def _make_distance_coord(metadata: XMLLaserZones):
+    """
+    Make the base coordinates from the metadata.
+    """
+    zones = metadata.zones
+    zone = next(iter(zones.values()))
+    dx = metadata.original_spatial_sampling_interval
+    distance = get_coord(
+        start=(zone.start_channel - 1) * dx,
+        stop=zone.end_channel * dx,
+        step=dx,
+        units=metadata.units,
+    )
+    return distance
+
+
+def _make_time_coord(file_start_times, metadata: XMLLaserZones):
+    """Create time coord for each file."""
+    dt = dc.to_timedelta64(1.0 / metadata.output_temporal_sampling_rate)
+    nt = metadata.number_of_frames
+    for start in file_start_times:
+        yield get_coord(start=start, stop=start + dt * nt, step=dt, units="s")
+
+
+def _make_base_attrs_dict(metadata: XMLLaserZones):
+    """
+    Make the base attributes and coordinates from metadata.
+    """
+    zones = metadata.zones
+    assert len(zones) == 1, "expecting single zone per metadata file."
+    zone_name = next(iter(zones))
+
+    ius = metadata.das_interrogator_serial
+    assert len(ius) == 1, "expecting one interrogator."
+    iu_name = next(iter(ius.values()))
+    attrs = dict(
+        pulse_width_ns=metadata.pulse_width_ns,
+        gauge_length=metadata.gauge_length_m,
+        instrument_id=iu_name,
+        zone_name=zone_name,
+    )
+    return attrs
+
+
+def _get_path_datetime_64(paths):
+    """Get a series of datetime64 and path as index."""
+    ser = pd.Series(x.name for x in paths)
+    split = ser.str.split("_", expand=True)
+    year = split[1].str[:4]
+    month = split[1].str[4:6]
+    day = split[1].str[6:8]
+    hour = split[1].str[9:11]
+    minute = split[1].str[11:13]
+    second = split[1].str[13:15]
+    frac_sec = split[2].str.split("Z", expand=True)[0]
+    iso8601 = (
+        year
+        + "-"
+        + month
+        + "-"
+        + day
+        + "T"
+        + hour
+        + ":"
+        + minute
+        + ":"
+        + second
+        + "."
+        + frac_sec
+    )
+    dt = dc.to_datetime64(iso8601.values)
+    return pd.Series(dt, index=ser.values)
+
+
+def _paths_to_df(paths, metadata, attr_cls):
+    """Convert paths to dataframe of info."""
+    attrs = _paths_to_attrs(paths=paths, metadata=metadata, attr_cls=attr_cls)
+    return _model_list_to_df(attrs)
+
+
+def _read_single_file(path, metadata, time, distance):
+    """Read a single file into a patch."""
+    assert not metadata.transposed_data, "Cant handle data transposition yet."
```

Comment:
Do we want to raise maybe a warning when the number of channels and/or time samples from the XML file is different from the raw binary data.shape? Currently, we read the data based on the XML file, which makes sense, but I think it doesn't hurt to raise a warning.

Reply:
I just added an assert on the shape being what we expect, I think that's probably the best we can do in this case.